HelgaShiryaeva/assignment2/BinaryTree.java

```diff
@@ -1,104 +1,166 @@
 package assignment2;
 
 import java.util.HashSet;
-import java.util.Optional;
+import java.util.NoSuchElementException;
 import java.util.Set;
 
 /**
  * This class represents a Binary tree data structure
  * @param <T> is a type of values stored in nodes of a tree
+ * All values stored in a binary tree are unique
  */
-public class BinaryTree <T> {
-    private Set<Node<T>> nodes;
-    private Node<T> root;
-
-    public BinaryTree(Node<T> root) {
-        this.root = root;
-        nodes = new HashSet<>();
-        nodes.add(root);
+public class BinaryTree<T> {
+    private Node root;
+    private Set<T> values;
+
+    public BinaryTree(T rootValue) {
+        this.root = new Node(rootValue);
+        values = new HashSet<>();
+        values.add(rootValue);
     }
 
     /**
-     * All the nodes of a tree should be inserted into a nodes set
-     * in order to avoid duplicates. A node which is not in the
-     * set of nodes is not considered to be a part of a binary tree.
-     * @param newNode is a node to be inserted in the nodes set
-     * @throws IllegalArgumentException if node is already in a binary tree or if
-     * the node with such value is already exists in the tree
-     */
-    public void addNode(Node<T> newNode) {
-        Optional<Node<T>> nodeOptional = nodes.stream()
-                .filter(e -> e.getValue().equals(newNode.getValue()))
-                .findFirst();
-        if (nodes.contains(newNode) || nodeOptional.isPresent()) {
-            throw new IllegalArgumentException("Node with value " + newNode.getValue() + " already exists in the tree");
-        }
-        nodes.add(newNode);
+     * This method adds a new node in a binary tree
+     * If the parent of a new node already has a child on
+     * the requested side an old child would be replaced by a new one
+     * @param value is a value of a new node
+     * @param parent is a parent of the node
+     * @param side on which side of parent a new node should be added
+     * */
+    public void add(T value, T parent, Side side) {

```

Comment:
Optional: if the class assumes that values are unique, it would be good to enforce it at the time when new element is added.